oac_summary.m4

```diff
@@ -0,0 +1,96 @@
+# of section followed by an underscore followed by a
+# shell-variable-ified representation of line.
+#
+# There are no restrictions on the contents of section and topic; they
+# can be variable references (although the use case for this is
+# dubious) and they can contain most ASCII characters (escape
+# characters excluded).  Note that some care must be taken with the
+# unique check and this liberal rule, as the unique check is after the
+# string has been run through AS_TR_SH.  Basically, any character that
+# is not legal in a shell variable name will be turned into an
+# underscore.  So the strings "Section_foo" and "Section-foo" would be
+# the same as far as the unique check is concerned.
+#
+# The input strings are evaluated during OAC_SUMMARY_ADD, not during
+# OAC_SUMMARY_PRINT.  This seems to meet the principle of least
+# astonishment.  A common pattern is to call
+# OAC_SUMMARY_ADD([Resource Type], [Component Name], [$results])
+# and then unset $results to avoid namespace pollution.  This will
+# work properly with the current behavior, but would result in odd
+# errors if we delayed evaluation.
+# -----------------------------------------------------------
+AC_DEFUN([OAC_SUMMARY_ADD],[
+    OAC_VAR_SCOPE_PUSH([oac_summary_line oac_summary_newline oac_summary_key])
+
+    # The end quote on the next line is intentional!
+    oac_summary_newline="
+"
+    oac_summary_line="$2: $3"
+    oac_summary_key="AS_TR_SH([$1])_AS_TR_SH([$2])"
+
+    # Use the section name variable as an indicator for whether or not
+    # the section has already been created.
+    AS_IF([AS_VAR_TEST_SET([oac_summary_section_]AS_TR_SH([$1])[_name])],
+          [],
+          [AS_VAR_SET([oac_summary_section_]AS_TR_SH([$1])[_name], ["$1"])
+           OAC_APPEND([oac_summary_sections], [AS_TR_SH([$1])])])
+
+    # Use the summary key as indicator if the section:topic has already
+    # been added to the results for the given section.
+    AS_IF([AS_VAR_TEST_SET([${oac_summary_key}])],
+          [],
+          [AS_VAR_SET([${oac_summary_key}], [1])
+           dnl this is a bit overcomplicated, but we are basically implementing
+           dnl a poor mans AS_VAR_APPEND with the ability to specify a separator,
+           dnl because we have a newline separator in the string.
+           AS_IF([AS_VAR_TEST_SET([oac_summary_section_]AS_TR_SH([$1])[_value])],
+                 [AS_VAR_APPEND([oac_summary_section_]AS_TR_SH([$1])[_value],
+                                ["${oac_summary_newline}${oac_summary_line}"])],
+                 [AS_VAR_SET([oac_summary_section_]AS_TR_SH([$1])[_value],
+                             ["${oac_summary_line}"])])])
+
+    OAC_VAR_SCOPE_POP
+])
+
+dnl If $1 is "stderr", then send the output to stderr.  Otherwise,
+dnl send the output to stdout.
+AC_DEFUN([OAC_SUMMARY_PRINT],[
+    OAC_VAR_SCOPE_PUSH([oac_summary_section oac_summary_section_name oac_sumary_fd])
+    AS_IF([test "$1" = "stderr"],
```

Comment:
I'd prefer do this at autoconf time; this shouldn't be a cofigure time option.  Something like:

```
m4_define([oac_summary_print_fd], [m4_default([$1], [1])])
echo "foobar" >&oac_summary_print_fd
```

Note that I didn't actually test that, so some adjustments might be needed.  I'd also claim that the default should probably actually be AS_MESSAGE_FD instead of 1, so that the summary doesn't print when run with `configure -q`, but that's a bigger discussion.

Reply:
You mean as a command line option? I much prefer this method where I can just hard-code it with the desired behavior.  There's only one of "me" over here, so it doesn't make a lot of sense to add complexity for what is, after all, a pretty simple thing.

Note that I also don't want to change the output stream for _all_ AC output - I just want the old behavior where the summary comes out on stderr and everything else on stdout.

Reply:
I looked at `m4sh` yesterday and didn't find anything that would have allowed me to do it at autogen time; I neglected to look at the `m4` macros. `m4_default` looks perfect -- thanks for the suggestion.

I wanted to keep the ability to pass `stderr` instead of `2` (it's just a little more expressive to read `OAC_SUMMARY_PRINT([stderr])` than `OAC_SUMMARY_PRINT([2])`), so I ended up going with:

```
m4_define([oac_summary_print_fd], [m4_default(m4_if([$1], [stderr], [2]), [1])])
```

Reply:
@rhc No, there's no difference to what you'll be doing.  It's still effectively hard-coded in `config/pmix_configure.m4`; it's just done slightly differently (more Autoconf-ish).

Reply:
Got it - thx! I realized it myself earlier this morning as I looked more closely at what you guys were doing.

Reply:
> I looked at `m4sh` yesterday and didn't find anything that would have allowed me to do it at autogen time; I neglected to look at the `m4` macros. `m4_default` looks perfect -- thanks for the suggestion.

Right; m4sh is all about making writing shell code easier, so everything in that section is really configure time code.  There are some weird exceptions, but for the most part, you need to look at the m4sugar section (https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/html_node/Programming-in-M4sugar.html) for autoconf-time macros.  In this case, it's all renames of m4 macros, but same deal.

> I wanted to keep the ability to pass `stderr` instead of `2` (it's just a little more expressive to read `OAC_SUMMARY_PRINT([stderr])` than `OAC_SUMMARY_PRINT([2])`), so I ended up going with:
> 
> ```
> m4_define([oac_summary_print_fd], [m4_default(m4_if([$1], [stderr], [2]), [AS_MESSAGE_FD])])
> ```

Ugh; I guess, but that's ugly.